rnd/autogpt_server/autogpt_server/executor/manager.py

```diff
@@ -402,62 +396,89 @@ class Executor:
     @classmethod
     def on_node_executor_start(cls):
         configure_logging()
-        cls.logger = logging.getLogger("node_executor")
+
         cls.loop = asyncio.new_event_loop()
+        cls.pid = os.getpid()
+
         cls.loop.run_until_complete(db.connect())
         cls.agent_server_client = get_agent_server_client()
 
+        # Set up shutdown handlers
+        cls.shutdown_lock = threading.Lock()
+        atexit.register(cls.on_node_executor_stop)  # handle regular shutdown
+        signal.signal(  # handle termination
+            signal.SIGTERM, lambda _, __: cls.on_node_executor_sigterm()
+        )
+
+    @classmethod
+    def on_node_executor_stop(cls):
+        if not cls.shutdown_lock.acquire(blocking=False):
```

Comment:
Why do we need lock? isn't there only one process & one thread instance executing it ?  Why can't it be a boolean flag variable?

If not, is it another process not another thread ? if it is another process, shouldn't we use multiprocess.Lock instead ?

Reply:
Once one of the cleanup handlers has started, there is no reason for the other to also run. The only thing that can possibly do is cause problems.

I started off with a boolean flag, iterated to get rid of race conditions and interrupt issues, and ended up with this. Although in this final iteration, `if not cls.shutdown: cls.shutdown = True` would probably also work.

Reply:
So those two handlers are running within the same thread serially, right ? so maybe we can try to just use boolean ?

Reply:
Is there a reason why we shouldn't be using a lock here? It's a cheap atomic operation and I'm just glad it finally works :)